### DISCUSSION TITLE
Question regarding scaling to 1B 256D float vectors + tuning

### DISCUSSION THREAD

**Discussion by vishal-kvn** [milvus-io/milvus#5121] [Q&A and General discussion] [answered]
Hello

We are considering using Milvus for an internal project that has upto 1 billion 256 dimension float vectors. For the proof of concept, I am using a dataset of 400 million vectors. I have a few questions regarding scaling and existing capabilities.

### Hardware Config
```
GPU type: Tesla V100
GPU memory: 16 GB
vCPUs: 64
Memory: 488 GB
Local Storage: 1TB
```

### Milvus Server Config
```
cache:
    cache_size: 256GB
    insert_buffer_size: 1GB
    preload_collection: ['demo_collection']
gpu:
    enable: true
    cache_size: 12GB
    gpu_search_threshold: 1001
```

For the proof of concept, I am using a GPU to build an IVF_PQ index and CPU to search for vectors. I have randomly chosen 10000 vectors and computed the ground truth for these vectors which I use for calculating recalls.  Here are my questions -

1. The index build time with `nlist=4096` is 40 minutes, which is fast. Does this compare to your previous benchmarks?
2. Does Milvus support a search by ids functionality? I could not find an example in the docs. Apologies in case I missed it.
3. In order to get around 2, I have been using `get_entity_by_id` to get a list of query vectors for a list of ids and then passing them to the search method. It takes about **50 seconds** to look up vectors for a batch of **1000 ids**. Is this expected? I understand that look ups are dependent on disk I/O but I am wondering if there is a way of making the look ups faster?
4. Another observation is that the recalls remain **constant** with increasing values of nprobe(`[16, 32, 64, 128, 256]`). I see an increase in search time as the nprobe increases but the recall remains the same. Did you come across this situation before?
5. To calculate nlist, I took the average row_count(1093460) from the TableFiles table and applied the formula(`nlist = 4 * sqrt(n)`) as per the docs to get nlist ~ 4096. Is this the right way to calculate nlist?
6. As per the sizing-tool(https://zilliz.com/sizing-tool), an IVF_PQ index for a 400M 256D float vector requires 398GB file size and 168GB memory. Based on my hardware config, shouldn’t the resources be sufficient?
7. Based on the above points, do you have any recommendations or best practices to speed up vector look ups and improve recalls?

Thanks in advance!

**Comment by yhmo** [comment, marked as answer]
1. Our test team has report for 1 billion 256 vectors, IVF_SQ8, nlist=16384, with 4 GPU devices(GeForce2080), totally it cost 4 hours to build index. Your index type is PQ, nlist is 4096. With our experience, nlist-16384 is much slower than nlist-4096. And V100 is powerful than GeForce2080. So I think 40 minutes is make sense.
2. Milvus doesn't support search by ids
3. There is an issue #4756 about performance of get_entity_by_id(). We make a improvement recently, in v1.1.0, which will be released today(May 7th). The improvement can reduce time cost to 1/4 of old version.
4. my colleague will explain this
5. correct
6. If all the segments have finish index, cache=256GB is sufficient.
7. The index_file_size(parameter of create_collection()) is 1024MB? For 1 billion 256 dimension float vectors, there will be 1000 segments on the disk. I think you can try index_file_size=2048, to reduce the count of segment files.

**Reply by shengjun1985** [reply]
4. There will be precision loss when using IVF_PQ query, so it is normal for recall rate to have upper limit.

**Reply by vishal-kvn** [reply]
Thank you for the response. It is very useful. Regarding `4.`, I increased m from 16 to 64 and the recall rate went from `.51` to `.96`.
